### PR TITLE
General node editor

### LINKED ISSUE #576: General Node-editor
Generalized node editor for particle and animation systems.

### PR DESCRIPTION
**Pull request by ludvigarlebrink**
resolves #576

### CODE REVIEW

File: src/Editor/Resources.hpp [hunk 1]
```diff
@@ -92,6 +111,24 @@ class ResourceList {
         
         /// The id of the next model to create.
```
Comment (NinjaDanz3r) [inline src/Editor/Resources.hpp]: MINOR Unused private function: 'ResourceList::GetInstance' [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)

File: src/Engine/Manager/RenderManager.cpp [hunk 2]
```diff
@@ -321,12 +322,22 @@ void RenderManager::Render(World& world, const glm::mat4& translationMatrix, con
...
+Component::AnimationController* RenderManager::CreateAnimation() {
+    return animationControllers.Create();
 }
 
-Component::Animation* RenderManager::CreateAnimation(const Json::Value& node) {
-    Component::Animation* animation = animations.Create();
+Component::AnimationController* RenderManager::CreateAnimation(const Json::Value& node) {
+    Component::AnimationController* animationController = animationControllers.Create();
     
     // Load values from Json node.
```
Comment (NinjaDanz3r) [inline src/Engine/Manager/RenderManager.cpp]: MINOR The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Engine/Manager/RenderManager.cpp [hunk 3]
```diff
@@ -321,12 +322,22 @@ void RenderManager::Render(World& world, const glm::mat4& translationMatrix, con
...
-Component::Animation* RenderManager::CreateAnimation() {
-    return animations.Create();
+void RenderManager::UpdateAnimations(float deltaTime) {
+    // Update all enabled animation controllers.
+    for (Component::AnimationController* animationController : animationControllers.GetAll()) {
+        if (animationController->IsKilled() || !animationController->entity->enabled)
+            continue;
+    
+        animationController->UpdateAnimation(deltaTime);
+    }
```
Comment (NinjaDanz3r) [inline src/Engine/Manager/RenderManager.cpp]: MINOR Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 14 issues
* ![MINOR][MINOR] 13 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L360](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L360): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L399](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L399): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L417](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L417): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L437](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L437): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L458](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L458): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L475](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/RenderManager.cpp#L475): The function 'GetControllers' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::rectangle' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/790ec483ff4f340210f0310eba85a3f01ba18857/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::cube' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 14 issues
* ![MINOR][MINOR] 13 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L360](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L360): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L399](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L399): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L417](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L417): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L437](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L437): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L458](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L458): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L475](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/RenderManager.cpp#L475): The function 'GetControllers' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::rectangle' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/431b2447787e1054cdb8a771caf5f9d73034d496/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::cube' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 14 issues
* ![MINOR][MINOR] 13 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L360](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L360): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L399](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L399): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L417](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L417): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L437](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L437): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L458](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L458): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L475](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/RenderManager.cpp#L475): The function 'GetControllers' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::rectangle' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
1. ![MINOR][MINOR] [ResourceManager.hpp#L28](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/49fb7615d97e8024a9c40d39371e871558528b3d/src/Engine/Manager/ResourceManager.hpp#L28): Member variable 'ResourceManager::cube' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by ludvigarlebrink** [comment]
Please don't review this yet.

File: src/Editor/GUI/Editors/EntityEditor.cpp [hunk 4]
```diff
@@ -170,21 +172,30 @@ void EntityEditor::SetVisible(bool visible) {
...
+    if (ImGui::Button("Select animation controller##Animation"))
+        ImGui::OpenPopup("Select animation Controller##Animation");
 
-    if (ImGui::BeginPopup("Select model##Animation")) {
-        ImGui::Text("Models");
+    if (ImGui::BeginPopup("Select animation controller##Animation")) {
+        ImGui::Text("Animation controller");
         ImGui::Separator();
+
+    //   for (Animation::AnimationController* controller : Resources().animationControllers) {
```
Comment (NinjaDanz3r) [inline src/Editor/GUI/Editors/EntityEditor.cpp]: Remove or refactor commented out code.

File: src/Editor/Resources_BACKUP_7056.cpp [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+#include <Engine/Manager/Managers.hpp>
+#include <Engine/Manager/ResourceManager.hpp>
+#include <Engine/Util/FileSystem.hpp>
+
+using namespace std;
+
+string ResourceList::Resource::GetName() const {
+    switch (type) {
+    case Type::SCENE:
+<<<<<<< HEAD
```
Comment (NinjaDanz3r) [inline src/Editor/Resources_BACKUP_7056.cpp]: Resolve this merge conflict.

File: src/Engine/Component/AnimationController.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            /**
+             * @return Vector with bones.
+             */
+            std::vector<glm::mat4> GetBones();
+
+            Animation::AnimationController* controller = nullptr;
+        private:
+
+    };
+}
```
Comment (NinjaDanz3r) [inline src/Engine/Component/AnimationController.hpp]: Add newline at end of file.

File: src/Engine/Util/Node.hpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        file->write(reinterpret_cast<char*>(name), 128);
+        file->write(reinterpret_cast<char*>(&numInputSlots), sizeof(uint32_t));
+    }
+
+    virtual void Load(std::ifstream* file) {
+        file->read(reinterpret_cast<char*>(&index), sizeof(uint32_t));
+        file->read(reinterpret_cast<char*>(name), 128);
+        file->read(reinterpret_cast<char*>(&numInputSlots), sizeof(uint32_t));
+    }
+};
```
Comment (NinjaDanz3r) [inline src/Engine/Util/Node.hpp]: Add newline at end of file.

**Comment by andens** [comment]
Closing as this is part of #677

**Pull request closed without merging**